### DISCUSSION TITLE
I'm using the same type with and without ID. How to handle this case?

### DISCUSSION THREAD

**Discussion by frederikhors** [urql-graphql/urql#1322] [Q&A] [answered]
I ran into a new problem.

I have:

```graphql
type Player {
  name: String!
}

type Query {
  players(search: String!): [Player!]
  friends(search: String!): [Player!]
}
```

In the `friends` query I do NOT return `id`s (because I don't need to cache them, these are `network-only` queries that I need to verify something).

Urql is (correctly) warning me about https://formidable.com/open-source/urql/docs/graphcache/errors/#15-invalid-key error.

I cannot use:

```js
const keys = {
  Player: () => null,
};
```

because I'm using `id`s for it elsewhere.

**How to handle this case?**

**Comment by kitten** [comment, marked as answer]
> I do NOT return ids (because I don't need to cache them, these are network-only queries that I need to verify something)

Still ask for the keyable field, in this case `id` then, I assume. Any query updates the cache, no matter what they mean for your app, so you'll have to consistently make sure that the same type will be keyable in the same manner.

**Reply by frederikhors** [reply]
Ok. I'm asking for `id` now but from the server the `id` is `0`.

Is it ok?

The error is gone now.

**Reply by frederikhors** [reply]
This is not OK. I cannot have multiple Players with the same ID (0).

**Reply by frederikhors** [reply]
The problem is I do not have IDs on server for `friends` special query.

**Reply by frederikhors** [reply]
I think this is an edge case.

**Reply by JoviDeCroock** [reply]
I mean, doesn't that mean that the backend data model is lacking since that would mean that your `Person.friend[0]` & `Person.friend[1]` are the same. In a normalised caching situation you'd prefer having the real ids for these people to ensure data consistency.

If it's really not possible you could _consider_ doing: `Player: (data) => data.id || null`

**Reply by kitten** [reply]
Yea, about the above, it seems like the schema is lacking consistency and information here. I'd assume that a `Player.friends` field could also return ID-less edges. But I'd assume that a `Player` entity would always need an ID.
